libraries/BLE/src/BLECharacteristicImp.cpp

```diff
@@ -0,0 +1,871 @@
+/*
+ * Copyright (c) 2015 Intel Corporation.  All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ */
+
+#include "BLEAttribute.h"
+#include "BLECharacteristicImp.h"
+
+#include "BLECallbacks.h"
+#include "BLEUtils.h"
+
+bt_uuid_16_t BLECharacteristicImp::_gatt_chrc_uuid = {BT_UUID_TYPE_16, BT_UUID_GATT_CHRC_VAL};
+bt_uuid_16_t BLECharacteristicImp::_gatt_ccc_uuid = {BT_UUID_TYPE_16, BT_UUID_GATT_CCC_VAL};
+
+BLECharacteristicImp::BLECharacteristicImp(const bt_uuid_t* uuid, 
+                                           unsigned char properties,
+                                           uint16_t handle,
+                                           const BLEDevice& bledevice):
+    BLEAttribute(uuid, BLETypeCharacteristic),
+    _value_length(0),
+    _value_buffer(NULL),
+    _value_updated(false),
+    _value_handle(handle),
+    _cccd_handle(0),
+    _attr_chrc_value(NULL),
+    _attr_cccd(NULL),
+    _subscribed(false),
+    _ble_device()
+{
+    _value_size = BLE_MAX_ATTR_DATA_LEN;// Set as MAX value. TODO: long read/write need to twist
+    _value = (unsigned char*)balloc(_value_size, NULL);
+    if (_value_size > BLE_MAX_ATTR_DATA_LEN)
+    {
+        _value_buffer = (unsigned char*)balloc(_value_size, NULL);
```

Comment:
Check `_value` and `_value_buffer` for NULL-ness. Heap is very small, it might well be NULL.